### PR TITLE
add Matrix & unit types, update generate mjml file to use Matrix type

### PR DESCRIPTION
**Pull request by hchiufaire**
Addresses #24.

- added types for px, %, and em, as well as `Matrix` which type-supports 1-4 values for css shorthands <br>(i.e.&nbsp;`padding: 0 10% 6em 4px;`)
  - updated MJML component script to use `Matrix` when parsing `{1,4}` from MJML types ([example from source](https://github.com/mjmlio/mjml/blob/master/packages/mjml-accordion/src/AccordionText.js#L20)), with appropriate units (px, %, em)
  - single value props are also changed to `Pixel`, `Percentage`, `Ephemeral` types where possible
- updated imports on script template and regenerated components
 
**Note:** typically the only unit-less number you can have as a CSS prop is "0", so the [new type definition](https://www.peerigon.com/en/blog/complete-guide-to-css-units-series-part-4-angles-time-dpi-unitless/#:~:text=Unitless%20values&text=The%20first%20is%20when%20setting,unitless%20number%20as%20a%20width.) of allowing any number without a unit is not as strict as it could be, but props like `flex` accept other unit-less numbers than "0", so this can be addressed later on in cases if needed

### Updated Tests
image

### New Type Checking in Web-Email-Server (after `yarn link`)
image
image

### CODE REVIEW

**Comment by IanEdington** [comment]
I was expecting this to affect more components.

Would you be able to write some tests for the use cases this covers?

Also, testing against our internal emails would be a good idea for this one to check if anything breaks.

**Review by IanEdington — COMMENTED** [review]
I was expecting this to affect more props.

Would you be able to write some tests for the use cases this covers?

Also, testing against our internal emails would be a good idea for this one to check if anything breaks.

File: scripts/generate-mjml-react.ts [hunk 1]
```diff
@@ -93,18 +95,21 @@ const ALLOW_ANY_PROPERTY = new Set(
 function getPropTypeFromMjmlAttributeType(
```
Comment (emmclaughlin) [inline scripts/generate-mjml-react.ts:95]: To go off @IanEdington point about testing, I am thinking we should consider refactoring this script into smaller testable functions. E.g. something like `getPropTypeFromMjmlAttributeType` could be it's own file that is tested

**Review by emmclaughlin — COMMENTED** [review]
This is looking good! Agree that we should add some tests for this. Tests on the script itself can probably be differed, but we could add a test on the Mjml elements that they accept matrix props, pixel props, and ephemeral props

File: test/mjml-component.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    expect(renderToMjml(button2)).toBe(
+      '<mj-button padding="0 1px">Button2</mj-button>'
+    );
+    expect(renderToMjml(button3)).toBe(
+      '<mj-button padding="0 1px 2%">Button3</mj-button>'
+    );
+    expect(renderToMjml(button4)).toBe(
+      '<mj-button padding="0 1px 2% 0px">Button4</mj-button>'
+    );
+  });
```
Comment (IanEdington) [inline test/mjml-component.spec.tsx:52]: These are great tests. Is it possible to write a negative test? Essentially we expect a type error to be thrown here?
No idea if this is possible in this format and I don't think it needs to be a blocker, but I know DefinitelyTyped has this kind of test: ie. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mjml-react/mjml-react-tests.tsx

**Review by IanEdington — COMMENTED** [review]
Looks good. I'll let @emmclaughlin have the final say on TS ect.

File: test/mjml-component.spec.tsx (hunk 2 above)
Comment (emmclaughlin) [reply test/mjml-component.spec.tsx:52]: I was thinking the same thing, but didn't think it should block this PR. I will add it as something to look into for the test plan

**Pull request merged**